File: framework/languages/dotnet/e2e/ResilienceLifecycle/Client/Scenarios/RlB6GrayFaultScenario.cs
// Verifies RL-B6 Gray Fault behavior.
using ResilienceLifecycle.Client.Support;
using ResilienceLifecycle.Shared;
using Zlink.HttpClient;

namespace ResilienceLifecycle.Client.Scenarios;

// RL-B6 verifies mixed gray failures, healthy-provider success, and recovery.
internal static class RlB6GrayFaultScenario
{
    public static async Task RunAsync(
        ZLinkHttpClient consumer,
        ZLinkHttpClient providerA,
        ZLinkHttpClient providerB)
    {
        await providerB.Post("/admin/fault/gray").AsyncRaw();

        var failures = 0;
        var successes = 0;
        for (var i = 0; i < 30; i++)
            try
            {
                var reply = (await consumer.Post("/profile/request")
                    .Body(new ProfileReq(i % 3 == 0 ? "gray" : "fast", $"rl-b6-{i}"))
                    .Async<ProfileRes>()).Body;
                if (reply.ProviderRid == "api-a") successes++;
            }
            catch
            {
                failures++;
            }

        ZlinkStreamAssert.Ensure(successes > 0 && failures > 0, "RL-B6 expected both healthy successes and gray failures.");
        await providerB.Post("/admin/fault/none").AsyncRaw();
        var followUp = (await consumer.Post("/profile/request")
            .Body(new ProfileReq("fast", "rl-b6-after"))
            .Async<ProfileRes>()).Body;
        ZlinkStreamAssert.Ensure(followUp.Value == "profile:fast", "RL-B6 follow-up request failed after clearing fault.");

        {
            using var timeout = new CancellationTokenSource(TimeSpan.FromSeconds(15));
            var waitA = providerA.Post("/evidence/wait").Body(new EvidenceWaitReq(["marker=rl-b6-"], []))
                .Async<string[]>(timeout.Token).AsTask();
            var waitB = providerB.Post("/evidence/wait").Body(new EvidenceWaitReq(["marker=rl-b6-"], []))
                .Async<string[]>(timeout.Token).AsTask();
            var completed = await Task.WhenAny(waitA, waitB);
            var evidence = (await completed).Body;
            timeout.Cancel();
            ZlinkStreamAssert.Ensure(evidence.Any(line => line.Contains("marker=rl-b6-", StringComparison.Ordinal)),
                "RL-B6 did not record expected evidence 'marker=rl-b6-'.");
        }
        {
            using var timeout = new CancellationTokenSource(TimeSpan.FromSeconds(15));
            var waitA = providerA.Post("/evidence/wait").Body(new EvidenceWaitReq(["marker=rl-b6-after"], []))
                .Async<string[]>(timeout.Token).AsTask();
            var waitB = providerB.Post("/evidence/wait").Body(new EvidenceWaitReq(["marker=rl-b6-after"], []))
                .Async<string[]>(timeout.Token).AsTask();
            var completed = await Task.WhenAny(waitA, waitB);
            var evidence = (await completed).Body;
            timeout.Cancel();
            ZlinkStreamAssert.Ensure(evidence.Any(line => line.Contains("marker=rl-b6-after", StringComparison.Ordinal)),
                "RL-B6 did not record expected evidence 'marker=rl-b6-after'.");
        }

        Console.WriteLine("scenario RL-B6 passed");
    }
}

File: framework/languages/node/e2e/ResilienceLifecycle/Client/Scenarios/rl-b6-gray-fault-scenario.ts
// RL-B6: 한 provider의 gray failure를 다른 replies와 격리한다 시나리오를 검증한다.
import type { ProfileRes } from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { postJson } from '../../../http-client';
import { profileReq } from '../Support/resilience-helpers';
import { ensure } from '../Support/scenario-assert';

export async function runRlB6(options: ClientOptions): Promise<void> {
  await postJson(options.providerBUrl, '/admin/fault/gray');

  let failures = 0;
  let healthySuccesses = 0;
  for (let i = 0; i < 60 && (failures === 0 || healthySuccesses === 0); i += 1) {
    try {
      const reply = await postJson<ProfileRes>(
        options.consumerUrl,
        '/profile/request/no-retry',
        profileReq(`rl-b6-${i}`)
      );
      if (reply.providerRid === 'api-a') {
        healthySuccesses += 1;
      }
    } catch {
      failures += 1;
    }
  }

  ensure(healthySuccesses > 0 && failures > 0, 'RL-B6 expected both healthy successes and gray failures.');

  await postJson(options.providerBUrl, '/admin/fault/none');
  const followUp = await postJson<ProfileRes>(
    options.consumerUrl,
    '/profile/request',
    { value: 'fast', marker: 'rl-b6-after' }
  );
  ensure(followUp.value === 'profile:fast', 'RL-B6 follow-up request failed after clearing fault.');

  const grayEvidence = await Promise.any([
    postJson<string[]>(options.providerAUrl, '/evidence/wait', { contains: 'marker=rl-b6-' }),
    postJson<string[]>(options.providerBUrl, '/evidence/wait', { contains: 'marker=rl-b6-' })
  ]);
  ensure(
    grayEvidence.some((line) => line.includes('marker=rl-b6-')),
    'RL-B6 did not record expected gray-failure evidence.'
  );

  const afterEvidence = await Promise.any([
    postJson<string[]>(options.providerAUrl, '/evidence/wait', { contains: 'marker=rl-b6-after' }),
    postJson<string[]>(options.providerBUrl, '/evidence/wait', { contains: 'marker=rl-b6-after' })
  ]);
  ensure(
    afterEvidence.some((line) => line.includes('marker=rl-b6-after')),
    'RL-B6 did not record expected recovery evidence.'
  );

  console.log('scenario RL-B6 passed');
}
